### PR TITLE
fix(owl-bot): only use lock around running post-processor

### LINKED ISSUE #3257: owlbot: owlbot: run label not removed
There is a race condition between the `synchronize` and `labeled` webhooks events, both of which attempt to kick off the owlbot post-processor.

If `synchronize` wins the race, the `owlbot: run` label will be left on a PR.

## Solution

We should remove the `owlbot: run` label on labeled events, whether we've been able to run the post-processor or not.

### PR DESCRIPTION
**Pull request by chingor13**
Currently, owlbot obtains a lock on a owner/repo/sha as soon as an event that could trigger a post-processor event runs. Instead, it should only attempt to run the post-processor under a lock.

This change moves the lock acquisition to a helper that wraps the run post-processor logic. This allows the webhook handlers to do their pre-checks and clean up no matter if it actually ends up running the post-processor or not.

Fixes #3257

### CODE REVIEW

File: packages/owl-bot/src/owl-bot.ts [hunk 1]
```diff
@@ -217,10 +189,10 @@ function OwlBot(privateKey: string | undefined, app: Probot, db?: Db): void {
           owner,
           repo,
           defaultBranch,
+          sha: context.payload.pull_request.head.sha,
         },
         context.octokit
       );
-      lock.release();
```
Comment (chingor13) [inline packages/owl-bot/src/owl-bot.ts:223]: Note: we also had an unawaited promise here

**Review by SurferJeffAtGoogle — CHANGES_REQUESTED** [review]

File: packages/owl-bot/src/owl-bot.ts [hunk 2]
```diff
@@ -433,6 +406,44 @@ function isBotAccount(sender: string): boolean {
...
+async function runPostProcessorWithLock(
+  appId: number,
+  privateKey: string,
+  project: string,
+  trigger: string,
+  opts: RunPostProcessorOpts,
+  octokit: Octokit,
+  breakLoop = true
+) {
+  // Short-circuit if post-processor already running for this SHA:
```
Comment (bcoe) [inline packages/owl-bot/src/owl-bot.ts:418]:
```suggestion
  // Short-circuit if post-processor already running for this SHA, preventing two post-processor images
  // from both executing and pushing changes against the same PR.
```

File: packages/owl-bot/src/owl-bot.ts [hunk 3]
```diff
@@ -433,6 +406,44 @@ function isBotAccount(sender: string): boolean {
   return /.*\[bot]$/.test(sender);
 }
 
+async function runPostProcessorWithLock(
```
[the remark this answers is not in the document]
Comment (bcoe) [reply packages/owl-bot/src/owl-bot.ts:409]: Now:

```js
  await removeOwlBotRunLabel(owner, repo, prNumber, octokit);
```

will execute, whereas prior to this change the lock was checked before calling.

**Review by bcoe — APPROVED** [review]
Looks reasonable to me once @SurferJeffAtGoogle's comments are addressed.

File: packages/owl-bot/src/owl-bot.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (chingor13) [reply packages/owl-bot/src/owl-bot.ts:409]: Yes, previously the `pull_request.labeled` handler tried to do everything under the lock. This prevents the bot from removing the `owlbot:run` label.

Only the actual running of the post-processor requires running under lock. We make it so both the label event handler and the code push event handler call the post-processor with the lock. Any other decision making on whether or not to run is localized in those handlers.

**Review by SurferJeffAtGoogle — APPROVED** [review]

**Pull request merged**